package eu.stratosphere.language.binding.protos;

option java_outer_classname = "StratospherePlan";

message ProtoPlan{

	enum ValueType{
		StringValue = 0;
		IntValue = 1;
		BooleanValue = 2;
		FloatValue = 3;
	}

	enum VertexType{
		Map = 0;
		Reduce = 1;
		Join = 2;
		Cross = 3;
		CoGroup = 4;
		Union = 5;
		TextInputFormat = 6;
		CsvInputFormat = 7;
		// Do we need this two in input and output here?
		DelimitedInputFormat = 8;
		FileInputFormat = 9;
		JDBCInputFormat = 10;
		CsvOutputFormat = 11;
		DelimitedOutputFormat = 12;
		FileOutputFormat = 13;
		BulkIteration = 14;
		DeltaIteration = 15;
	}

	message KeyValuePair{
		required string key = 1;
		required string value = 2;
	}

	message Vertex{
		required VertexType type = 1;
		// Save the ID's of input vertices in int-arrays.
		// It's an array to save the IDs of multiple inputs for Unioning
		repeated int32 inputs1 = 2[packed = true];
		// The second array is used for operators with two inputs only (CoGroup, Cross, Join)
		repeated int32 inputs2 = 3[packed = true];
		repeated ValueType outputTypes = 4;
		repeated KeyValuePair params = 5;
		required int32 ID = 6;
	}

	repeated Vertex vertices = 1;
	repeated int32 DataSinkIDs = 2;
}